safehttp/mux.go

```diff
@@ -0,0 +1,144 @@
+//
+// Patterns names are fixed, rooted paths, like "/favicon.ico", or rooted
+// subtrees like "/images/" (note the trailing slash). Longer patterns take
+// precedence over shorter ones, so that if there are handlers registered for
+// both "/images/" and "/images/thumbnails/", the latter handler will be called
+// for paths beginning "/images/thumbnails/" and the former will receive
+// requests for any other paths in the "/images/" subtree.
+//
+// Note that since a pattern ending in a slash names a rooted subtree, the
+// pattern "/" matches all paths not matched by other registered patterns,
+// not just the URL with Path == "/".
+//
+// If a subtree has been registered and a request is received naming the subtree root without its trailing slash, ServeMux redirects that request to
+// the subtree root (adding the trailing slash). This behavior can be overridden
+// with a separate registration for the path without the trailing slash. For
+// example, registering "/images/" causes ServeMux to redirect a request for
+// "/images" to "/images/", unless "/images" has been registered separately.
+//
+// Patterns may optionally begin with a host name, restricting matches to URLs
+// on that host only. Host-specific patterns take precedence over general
+// patterns, so that a handler might register for the two patterns "/codesearch"
+// and "codesearch.google.com/" without also taking over requests for
+// "http://www.google.com/".
+//
+// ServeMux also takes care of sanitizing the URL request path and the Host
+// header, stripping the port number and redirecting any request containing . or
+// .. elements or repeated slashes to an equivalent, cleaner URL.
+//
+// Multiple HTTP methods can be served for one pattern and the user is expected
+// to register a handler for each method supported. These will be combined
+// into one handler per pattern. The framework will then match each
+// incoming request to its underlying handler according to its HTTP method.
+type ServeMux struct {
+	mux     *http.ServeMux
+	domains map[string]bool
+	disp    Dispatcher
+
+	// Maps user-provided patterns to combined handlers which encapsulate
+	// multiple handlers, each one associated with an HTTP method.
+	handlers map[string]methodHandler
+}
+
+// NewServeMux allocates and returns a new ServeMux
+// TODO(@mattiasgrenfeldt, @mihalimara22): make domains a variadic of string ..//**literals**.
+func NewServeMux(dispatcher Dispatcher, domains ...string) *ServeMux {
+	d := map[string]bool{}
+	for _, host := range domains {
+		d[host] = true
+	}
+	return &ServeMux{
+		mux:      http.NewServeMux(),
+		domains:  d,
+		disp:     dispatcher,
+		handlers: map[string]methodHandler{},
+	}
+}
+
+// Handle registers a handler for the given pattern and method. If another
+// handler is already registered for the same pattern and method, Handle panics.
+func (m *ServeMux) Handle(pattern string, method string, h Handler) {
+	ch, ok := m.handlers[pattern]
```

Comment:
nit: please go through this PR and make sure the short names still make sense. `ch` is confusing since it holds a `methodHandler` now.